avoid error when sort mixture keys in symbol and string

y00rb opened the pull request:
when no route matches or missing required parameter, generate method will raise error.
but if mixture symbol and string in parameter, the sort method will
error:

``` bash
 comparison of Array with Array failed
```

the reason is because the sort method used <=> and thatn't work when
comparing a symble with at string:

``` bash
>> :foo <=> 'foo'
=> nil
```

this is fix #19314 in actionpack/lib/action_dispatch/journey/formatter.rb

rafaelfranca commented on actionpack/lib/action_dispatch/journey/formatter.rb:
```diff
@@ -39,7 +39,9 @@ def generate(name, options, path_parameters, parameterize = nil)
 
-        message = "No route matches #{Hash[constraints.sort].inspect}"
+        # Merge duplicate parameters which duplicated with different formart, and make all keys as symbol
+        primarty_parameters = constraints.inject({}){|h,(k,v)| h[k.to_sym] = v; h}
```
can we make them all strings?

y00rb replied:
make all string, them we can used `with_indifferent_access` method.

y00rb commented:
@rafaelfranca 
Thanks for your suggestion. it is fixed now. I also changed them all string in file `actionpack/lib/action_dispatch/routing/route_set.rb` that avoid same issues when sort.

rafaelfranca commented:
While this works it may make users confused. The reason they are getting this error is because they are merging user parameters with a hash what is discouraged as @pixeltrix explained at https://github.com/rails/rails/issues/19314#issuecomment-78902663 and merging string keys and symbols keys will just not show the real problem

What if, we just stop to sorting the hash?

y00rb commented:
yeah, for `"foo" <=> :foo'  case, i think most saturation are user send the wrong key.
such as putting "action" or "id" on *_path/_url method. in this way, the `:action, "action"` being same value.
and the error is cause by parameter missing.
how about adding duplicated parameters on error? same key with different format(symbol, string).

rafaelfranca commented:
Doesn't removing the `sort` call fix it?

y00rb commented:
yes, filter duplicated parameter(string format) on error. such as 
1. compare which is duplicated keys and format as string.
2. append message after "missing_keys"
```ruby
message << " checking duplicated: #{Hash[mixed_parameters.sort].inspect}" unless mixed_parameter.empty?
```

rafaelfranca commented:
Seems good

y00rb commented:
I will make the changes following discussion today.
then you can review it at tomorrow.

y00rb committed:
sort_by instead of sort

it is avoid sort errot within different and mixed keys.
used `sort_by` + `block` to list parameter by keys.
keep minimum changes

y00rb commented:
@rafaelfranca 
make minimum changes, replace sort via sort_by. event keys duplicated as string and symbol, just putting them in sorted list.

rafaelfranca commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

rafaelfranca commented:
Backported as 312f086
